pioreactor/actions/od_calibration_from_standards.py

```diff
@@ -0,0 +1,583 @@
+
+    click.confirm(
+        f"Confirm using channel {signal_channel} with angle {config['od_config.photodiode_channel'][signal_channel]}° position in the Pioreactor",
+        abort=True,
+        default=True,
+    )
+    angle = str(config["od_config.photodiode_channel"][signal_channel])
+    return name, angle, signal_channel
+
+
+def setup_HDC_instructions() -> None:
+    click.clear()
+    click.echo(
+        """ Setting up:
+    1. Place first standard into Pioreactor, with a stir bar.
+"""
+    )
+
+
+def start_stirring():
+    while not click.confirm("Reading to start stirring?", default=True):
+        pass
+
+    click.echo("Starting stirring.")
+
+    st = stirring(
+        target_rpm=config.getfloat("stirring", "target_rpm"),
+        unit=get_unit_name(),
+        experiment=get_latest_testing_experiment_name(),
+    )
+    st.block_until_rpm_is_close_to_target(abs_tolerance=120)
+    return st
+
+
+def plot_data(
+    x,
+    y,
+    title,
+    x_min=None,
+    x_max=None,
+    interpolation_curve=None,
+    highlight_recent_point=True,
+):
+    import plotext as plt  # type: ignore
+
+    plt.clf()
+
+    plt.scatter(x, y, marker="hd")
+
+    if highlight_recent_point:
+        plt.scatter([x[-1]], [y[-1]], color=204, marker="hd")
+
+    plt.theme("pro")
+    plt.title(title)
+    plt.plot_size(105, 22)
+
+    if interpolation_curve:
+        plt.plot(x, [interpolation_curve(x_) for x_ in x], color=204)
+        plt.plot_size(145, 42)
+
+    plt.xlim(x_min, x_max)
+    plt.show()
+
+
+def start_recording_standards(st: Stirrer, signal_channel):
+    target_rpm = st.target_rpm
+    voltages = []
+    od600_values = []
+    click.echo("Starting OD recordings.")
+
+    with start_od_reading(
+        cast(pt.PdAngleOrREF, config.get("od_config.photodiode_channel", "1")),
+        cast(pt.PdAngleOrREF, config.get("od_config.photodiode_channel", "2")),
+        interval=None,
+        unit=get_unit_name(),
+        fake_data=is_testing_env(),
+        experiment=get_latest_testing_experiment_name(),
+        use_calibration=False,
+    ) as od_reader:
+
+        def get_voltage_from_adc() -> float:
+            od_readings1 = od_reader.record_from_adc()
+            od_readings2 = od_reader.record_from_adc()
+            return 0.5 * (
+                od_readings1.ods[signal_channel].od
+                + od_readings2.ods[signal_channel].od
+            )
+
+        for _ in range(4):
+            # warm up
+            od_reader.record_from_adc()
+
+    while True:
+        click.echo("Recording next standard.")
+        standard_od = click.prompt("Enter OD600 measurement", type=float)
+        for i in range(4):
+            click.echo(".", nl=False)
+            sleep(0.5)
+
+        click.echo(".", nl=False)
+        voltage = get_voltage_from_adc()
+        click.echo(".", nl=False)
+
+        od600_values.append(standard_od)
+        voltages.append(voltage)
+
+        st.set_target_rpm(0)
```

Comment:
An option that may be more responsive: `st.set_state("sleeping")` to pause stirring, and `st.set_state("ready")` to restart stirring at the correct duty cycle.

Reply:
This is better, thanks - this should keep block_until_rpm_is_close_to_target still, though, right? on_sleeping_to_ready doesn't block, I think...

Reply:
I think keeping `block_until_rpm_is_close_to_target` is a good idea, yea. `on_sleeping_to_ready` doesn't block